File: tools/uicheck.js
#!/usr/bin/env node
/*
 * @gate The UI checker finds a frame drawn outside its parent, and stays quiet when nothing is
 *
 * Runs ui/UICheck.lua against a mocked WoW API with FABRICATED geometry.
 *
 * What this gate does NOT do is check the UI. It cannot: the harness has no font metrics and
 * does no layout, which is the entire reason /valuate uicheck exists. GetTop and GetBottom
 * return resolved screen coordinates and the harness has nothing to resolve them against.
 *
 * What it checks is the CHECKER - that when something is drawn outside its parent the report
 * says so and names it, and that when nothing is, the report is silent. A diagnostic that
 * cannot fail is worse than none, because a clean run from it would be read as evidence.
 *
 * The rectangles below are made up. They are the shape of the v0.158.0a defect - a row whose
 * text wrapped past the bottom edge of the row - written as coordinates, so the logic that
 * would have caught it can be proven here even though the measurement cannot.
 *
 * Usage:  node tools/uicheck.js
 */
"use strict";

const { load } = require("./luaharness.js");

const run = load(["ui/Shared.lua", "ui/UICheck.lua"]);

run(
  `
local failures, checks = {}, 0
local function ok(cond, what) checks = checks + 1 if not cond then table.insert(failures, what) end end
local function eq(got, want, what)
    checks = checks + 1
    if got ~= want then
        table.insert(failures, what .. " (got " .. tostring(got) .. ", wanted " .. tostring(want) .. ")")
    end
end

local ns = __ns

-- Fabricated geometry. The harness does no layout, so these are set by hand - which is the
-- honest version of what this gate can claim: the arithmetic and the reporting are tested,
-- the measuring is not, because there is nothing here to measure.
local function place(f, top, bottom, left, right)
    f.GetTop = function() return top end
    f.GetBottom = function() return bottom end
    f.GetLeft = function() return left end
    f.GetRight = function() return right end
    return f
end

UIParent = CreateFrame("Frame")
place(UIParent, 768, 0, 0, 1024)

local function newWindow()
    local root = CreateFrame("Frame", "ValuateUIFrame")
    place(root, 700, 100, 100, 900)
    root:Show()
    ns.ValuateUIFrame = root
    return root
end

-- ---- a clean window reports clean -------------------------------------------------------
-- The direction that matters most. If this file only proved the checker can complain, a
-- checker that complained about everything would pass it.
local root = newWindow()
local panel = CreateFrame("Frame", nil, root)
place(panel, 690, 110, 110, 890)
panel:Show()
local label = panel:CreateFontString(nil, "OVERLAY")
label:SetText("A line that fits")
place(label, 680, 660, 120, 400)
label:Show()

local problems, examined = ns.RunUICheck()
eq(problems, 0, "a window whose contents all fit reports nothing")
ok(examined > 0, "and says how much it looked at (" .. tostring(examined) .. ")")

-- ---- text spilling out of the bottom of its row -----------------------------------------
-- This is v0.158.0a exactly: a fixed-height row, and a detail that wrapped to three lines.
-- Every headless gate passed. It was visible the moment anyone opened the tab.
root = newWindow()
local row = CreateFrame("Frame", nil, root)
place(row, 600, 560, 110, 890)      -- a 40px row
row:Show()
local detail = row:CreateFontString(nil, "OVERLAY")
detail:SetText("No stat priority was ever published for that spec, so they were read off it")
place(detail, 596, 520, 120, 880)   -- wraps 40px past the bottom
detail:Show()

problems = ns.RunUICheck()
ok(problems >= 1, "text drawn below its own row is reported")

-- ---- and it names the thing ---------------------------------------------------------------
-- "1 problem found" is not actionable in a window with several hundred frames. The text of
-- the offending string is the fastest way to find it, so the report carries it.
__printed = {}
ns.RunUICheck()
local said = table.concat(__printed, " ")
ok(said:find("spills", 1, true) ~= nil, "the report says what went wrong")
ok(said:find("below", 1, true) ~= nil, "and in which direction")
ok(said:find("No stat priority", 1, true) ~= nil, "and quotes the text, so it can be found")

-- ---- the other three directions ------------------------------------------------------------
for _, case in ipairs({
    { name = "above", t = 700, b = 580 },
    { name = "left of", l = 90, r = 800 },
    { name = "right of", l = 120, r = 990 },
}) do
    root = newWindow()
    local box = CreateFrame("Frame", nil, root)
    place(box, 600, 560, 110, 890)
    box:Show()
    local fs = box:CreateFontString(nil, "OVERLAY")
    fs:SetText("out")
    place(fs, case.t or 590, case.b or 570, case.l or 120, case.r or 880)
    fs:Show()
    __printed = {}
    ns.RunUICheck()
    ok(table.concat(__printed, " "):find(case.name, 1, true) ~= nil,
       "a region past the " .. case.name .. " edge is reported")
end

-- ---- hidden things have no position, and that is not a defect ------------------------------
-- Pooled rows sit hidden with stale coordinates. Treating "no position" as a failure would
-- make every panel in the addon report dozens of problems it does not have.
--
-- The HIDDEN thing is the child, inside a SHOWN parent - which is the real shape of a frame
-- pool. Written the other way round first, with the parent hidden too, and the case passed
-- whether the child was checked or not: the parent's own guard was doing all the work, so
-- the assertion proved nothing about the one it was written for.
root = newWindow()
local list = CreateFrame("Frame", nil, root)
place(list, 600, 560, 110, 890)
list:Show()
local hiddenRow = CreateFrame("Frame", nil, list)
place(hiddenRow, 400, 300, 120, 880)   -- stale coordinates, far outside the list
hiddenRow:Hide()
eq(ns.RunUICheck(), 0, "a hidden row is not measured, however wrong its old coordinates are")

-- ---- an unresolved position is silence, not a failure ---------------------------------------
root = newWindow()
local floating = CreateFrame("Frame", nil, root)
floating:Show()
floating.GetTop = function() return nil end
eq(ns.RunUICheck(), 0, "a frame the client cannot place yet is skipped rather than blamed")

-- ---- quiet mode does not OPEN the window it was asked to inspect ------------------------------
-- /valuate selfverify runs everything the addon can judge on its own, and wants a verdict
-- rather than a report. A diagnostic that opens the window in order to measure it is measuring
-- something you did not have - and it would do that in the middle of a command that is
-- supposed to be passive.
root = newWindow()
root:Hide()
local opened = 0
Valuate.ShowUI = function() opened = opened + 1 end

local quietProblems = ns.RunUICheck(true)
eq(quietProblems, nil, "with the window closed, quiet mode returns no verdict at all")
eq(opened, 0, "and does not open the window to manufacture one")

-- The command form still does, because a person asking for it wants an answer.
ns.RunUICheck()
eq(opened, 1, "the command form opens it, because you asked")
Valuate.ShowUI = nil

-- ---- exactly one tab is marked ---------------------------------------------------------------
root = newWindow()
local function tab(lit)
    local b = CreateFrame("Button", nil, root)
    place(b, 120, 100, 200, 300)
    b:Show()
    b.accent = b:CreateTexture(nil, "OVERLAY")
    place(b.accent, 120, 118, 202, 298)
    if lit then b.accent:Show() else b.accent:Hide() end
    return b
end
root.tabs = { buttons = { a = tab(true), b = tab(false), c = tab(false) } }
eq(ns.RunUICheck(), 0, "one tab marked out of three is correct")

root.tabs = { buttons = { a = tab(true), b = tab(true), c = tab(false) } }
__printed = {}
ns.RunUICheck()
ok(table.concat(__printed, " "):find("at once", 1, true) ~= nil,
   "two tabs marked at once is reported - a bar on every tab marks nothing")

root.tabs = { buttons = { a = tab(false), b = tab(false) } }
__printed = {}
ns.RunUICheck()
ok(table.concat(__printed, " "):find("none of the", 1, true) ~= nil,
   "no tab marked at all is reported - which is what four of six did until v0.156.0a")

-- ---- what the check actually COVERED -------------------------------------------------------------
-- The walk measures only visible things, so it is a verdict on the tab you are looking at and
-- never on the window. "Clean, 412 things measured" read as the latter: someone could check from
-- the Scales tab and believe the Enhance tab they had never opened was fine. The number was true
-- and the impression it gave was not.
--
-- A FRESH ROOT, because everything above deliberately builds broken layout and RunUICheck walks
-- the whole tree. Reusing that root meant this block always took the FAILING branch, and the
-- mutation gutting the clean one survived - the assertion was reading a nearly identical
-- sentence from the other side of the if.
local cleanRoot = CreateFrame("Frame")
cleanRoot:SetWidth(800)
cleanRoot:SetHeight(600)
cleanRoot:Show()
ns.ValuateUIFrame = cleanRoot

local panels = {}
for _, name in ipairs({ "scales", "settings", "enhance", "todo", "about" }) do
    local pane = CreateFrame("Frame", nil, cleanRoot)
    pane:SetWidth(700)
    pane:SetHeight(500)
    pane:Hide()
    panels[name] = pane
end
panels.enhance:Show()

cleanRoot.tabs = {
    scalesPanel = panels.scales, settingsPanel = panels.settings,
    enhancePanel = panels.enhance, todoPanel = panels.todo, aboutPanel = panels.about,
    -- The table also carries non-panel entries, which must not be counted as tabs.
    selectTab = function() end,
    buttons = { (function()
        local btn = CreateFrame("Button", nil, cleanRoot)
        btn:SetWidth(90) btn:SetHeight(24)
        btn.accent = btn:CreateTexture(nil, "OVERLAY")
        btn.accent:SetWidth(90) btn.accent:SetHeight(2)
        btn.accent:Show()
        return btn
    end)() },
}

local active, total, unchecked = ns.UICheckCoverage()
eq(active, "enhance", "the visible panel is named, so the report says which tab it judged")
eq(total, 5, "the non-panel entries in the tabs table are not counted as tabs")
eq(unchecked, 4, "and the ones that were never open are counted")

__printed = {}
ns.RunUICheck()
local said = table.concat(__printed, "\\n")
ok(said:find("clean", 1, true) ~= nil,
   "the fixture produces a clean run, or the assertion below would prove the wrong branch")
ok(said:find("enhance tab", 1, true) ~= nil, "the summary names the tab it measured")
ok(said:find("open each one and run this again", 1, true) ~= nil,
   "and a CLEAN result still says how many tabs it did not measure")

-- No tabs table at all - an older core, or a window built by something else. It must not claim
-- coverage it cannot describe, and must not error.
cleanRoot.tabs = nil
active, total, unchecked = ns.UICheckCoverage()
eq(active, nil, "with no tabs table there is no tab to name")
eq(unchecked, 0, "and nothing is claimed about what was missed")
__printed = {}
ok(pcall(ns.RunUICheck), "and the check still runs rather than erroring")

return failures, checks
`,
  "uicheck",
  "the in-client UI checker"
);

File: ui/UICheck.lua
-- ui/UICheck.lua
-- /valuate uicheck - the questions only the running client can answer.
--
-- Sixty-six headless gates prove a great deal about this UI: that every tab is built the same
-- way, that buttons respond to the mouse, that rows are reused rather than leaked. What they
-- cannot prove is anything about PIXELS. The harness has no font metrics and does no layout;
-- its GetStringHeight is a rough model that says so in its own comment, and its GetBottom does
-- not exist at all, because there is nothing to resolve a position against.
--
-- So the defects that survive every gate are the ones with a coordinate in them:
--
--   * text drawn outside the frame that owns it - shipped in v0.158.0a, when a to-do row was
--     a fixed 40px and the detail on a guessed scale wrapped to three lines;
--   * a tab row grown wide enough to slide under the tab at the other end;
--   * anything pushed off the edge of the screen.
--
-- Each of those is obvious in half a second of looking and invisible to everything else. This
-- walks the real window in the real client and asks about them directly, using GetTop and
-- GetBottom - which return resolved screen coordinates, and so answer exactly the question
-- the harness cannot.
--
-- It is a DIAGNOSTIC, not a gate: it reports and returns counts. Nothing acts on the result.

local _, ns = ...

-- A pixel of slack. Anchors land on fractional coordinates and a font string sitting exactly
-- on its parent's edge is correct, not broken - flagging that would make this cry wolf on
-- every panel in the addon, which is the fastest way to make a diagnostic worthless.
local SLACK = 1

-- Depth cap. The window nests about six deep; anything past this is a cycle, and a diagnostic
-- that hangs the client is worse than no diagnostic.
local MAX_DEPTH = 12

-- Resolved geometry, or nil.
--
-- GetTop and friends return nil for a frame that is hidden or has no anchor the client can
-- resolve yet. That is not a defect - a hidden panel legitimately has no position - so every
-- check here treats "no coordinates" as "nothing to say" rather than as a failure.
local function Rect(f)
    -- All four, not just one. In the client they arrive together; in a mock they need not,
    -- and a checker that assumes the rest exist because the first did is a checker that
    -- errors instead of reporting.
    if not f then return nil end
    if not (f.GetTop and f.GetBottom and f.GetLeft and f.GetRight) then return nil end
    local top, bottom, left, right = f:GetTop(), f:GetBottom(), f:GetLeft(), f:GetRight()
    if not (top and bottom and left and right) then return nil end
    return { top = top, bottom = bottom, left = left, right = right }
end

local function Describe(region, parent)
    if region.GetText and region:GetText() then
        local text = region:GetText():gsub("|c%x%x%x%x%x%x%x%x", ""):gsub("|r", "")
        text = text:gsub("\n", " ")
        if #text > 44 then text = text:sub(1, 41) .. "..." end
        if text ~= "" then return '"' .. text .. '"' end
    end
    local name = region.GetName and region:GetName()
    if name then return name end
    local pname = parent and parent.GetName and parent:GetName()
    return (region.GetObjectType and region:GetObjectType() or "region") ..
        (pname and (" in " .. pname) or "")
end

-- Walks frames and their regions, calling visit(region, parent, isRegion).
local function Walk(frame, visit, depth, seen)
    depth = depth or 0
    seen = seen or {}
    if not frame or depth > MAX_DEPTH or seen[frame] then return end
    seen[frame] = true

    if frame.GetRegions then
        local regions = { frame:GetRegions() }
        for _, r in ipairs(regions) do
            if r then visit(r, frame, true) end
        end
    end
    if frame.GetChildren then
        local kids = { frame:GetChildren() }
        for _, c in ipairs(kids) do
            if c then
                visit(c, frame, false)
                Walk(c, visit, depth + 1, seen)
            end
        end
    end
end

-- quiet: measure and return, printing nothing.
--
-- For /valuate selfverify, which runs everything the addon can judge on its own and wants a
-- verdict rather than a report. It also must NOT open the window as the command form does -
-- a diagnostic that changes what is on your screen in order to inspect it is measuring
-- something you did not have.
function ns.RunUICheck(quiet)
    if not ns.ValuateUIFrame or not ns.ValuateUIFrame:IsShown() then
        if quiet then return nil, 0 end
        if Valuate.ShowUI then Valuate:ShowUI() end
    end
    local root = ns.ValuateUIFrame
    if not root then
        print("|cFFFF0000[Valuate]|r The window is not built, so there is nothing to check.")
        return 0, 0
    end

    local problems, examined = {}, 0
    local function report(what)
        problems[#problems + 1] = what
    end

    -- ---- 1. nothing is drawn outside the frame that owns it -----------------------------
    --
    -- The one that shipped. A fixed-height row and a detail line that wrapped to three put
    -- the second and third lines over whatever came next - correct in every gate, wrong on
    -- screen, and visible instantly to anyone who opened the tab.
    --
    -- Only VISIBLE things, and only where both rectangles resolve. A hidden row in a frame
    -- pool has no position and is not a defect.
    Walk(root, function(child, parent, isRegion)
        if not (child.IsShown and child:IsShown()) then return end
        if not (parent.IsShown and parent:IsShown()) then return end
        local c, p = Rect(child), Rect(parent)
        if not (c and p) then return end
        examined = examined + 1

        local over = {}
        if c.bottom < p.bottom - SLACK then over[#over + 1] = "below" end
        if c.top > p.top + SLACK then over[#over + 1] = "above" end
        if c.left < p.left - SLACK then over[#over + 1] = "left of" end
        if c.right > p.right + SLACK then over[#over + 1] = "right of" end
        if #over > 0 then
            report(string.format("%s spills %s its parent (%s)",
                isRegion and "text/texture" or "frame",
                table.concat(over, " and "), Describe(child, parent)))
        end
    end)

    -- ---- 2. nothing is off the screen ---------------------------------------------------
    --
    -- The window is clamped to the screen, but its CONTENTS are not: a panel that grew past
    -- the window, or a tab row that outgrew its row, ends up somewhere nobody can click.
    local screen = Rect(UIParent)
    if screen then
        Walk(root, function(child, _, isRegion)
            if isRegion then return end
            if not (child.IsShown and child:IsShown()) then return end
            local c = Rect(child)
            if not c then return end
            if c.right < screen.left or c.left > screen.right
               or c.top < screen.bottom or c.bottom > screen.top then
                report("a frame is off the edge of the screen (" .. Describe(child, nil) .. ")")
            end
        end)
    end

    -- ---- 3. the tab you are on is marked, and only that one -----------------------------
    --
    -- Four of six tabs had no accent texture at all until v0.156.0a, and the gate that now
    -- catches that can only see whether the texture EXISTS and what Show/Hide was called on
    -- it. Whether it is actually drawn - the draw layer, the anchor, the parent's strata -
    -- is a client question.
    local tabs = root.tabs and root.tabs.buttons
    if tabs then
        local lit, total = 0, 0
        for _, btn in pairs(tabs) do
            total = total + 1
            if btn.accent and btn.accent.IsShown and btn.accent:IsShown() then
                lit = lit + 1
                local a, b = Rect(btn.accent), Rect(btn)
                if a and b and (a.left < b.left - SLACK or a.right > b.right + SLACK) then
                    report("a tab's accent bar is wider than the tab it marks")
                end
            end
        end
        examined = examined + total
        if lit == 0 then
            report(string.format("none of the %d tabs is marked as current", total))
        elseif lit > 1 then
            report(string.format("%d of %d tabs are marked as current at once", lit, total))
        end
    end

    -- ---- 4. every button answers the mouse ----------------------------------------------
    --
    -- Same shape as the gate that found fourteen of these, but asked of the live frames: a
    -- HIGHLIGHT-layer texture is drawn by the client and needs no script, so a button has
    -- feedback if it has one of those OR a hover handler. Both are legitimate here.
    Walk(root, function(child, _, isRegion)
        if isRegion then return end
        if not (child.GetObjectType and child:GetObjectType() == "Button") then return end
        if not (child.IsShown and child:IsShown()) then return end
        if not child.GetBackdrop or not child:GetBackdrop() then return end
        examined = examined + 1

        if child.GetScript and child:GetScript("OnEnter") then return end
        local regions = { child:GetRegions() }
        for _, r in ipairs(regions) do
            if r and r.GetDrawLayer and r:GetDrawLayer() == "HIGHLIGHT" then return end
        end
        report("a button gives no sign it can be clicked (" .. Describe(child, nil) .. ")")
    end)

    -- ---- the report ----------------------------------------------------------------------
    if quiet then return #problems, examined end
    if #problems == 0 then
        print(string.format(
            "|cFF00FF00[Valuate]|r UI check: |cFF00FF00clean|r - %d things measured against the real client.",
            examined))
        print("  |cFFAAAAAAThis is the half the headless gates cannot see: real fonts, real anchors, " ..
              "real positions.|r")
    else
        print(string.format("|cFFFF8800[Valuate]|r UI check: %d problem%s in %d things measured.",
            #problems, #problems == 1 and "" or "s", examined))
        -- Deduplicated: a pooled row that is wrong is wrong in every copy, and twenty
        -- identical lines say nothing the first one did not.
        local seen, shown = {}, 0
        for _, p in ipairs(problems) do
            if not seen[p] then
                seen[p] = true
                shown = shown + 1
                if shown <= 12 then print("  |cFFFF8800-|r " .. p) end
            end
        end
        if shown > 12 then
            print(string.format("  |cFFAAAAAA...and %d more distinct.|r", shown - 12))
        end
    end

    return #problems, examined
end
